libs/db/badgerdb/db.go

```diff
@@ -0,0 +1,288 @@
+			val = []byte{}
+		}
+		return err
+	})
+	return val, err
+}
+
+func (b *BadgerDB) Has(key []byte) (bool, error) {
+	if len(key) == 0 {
+		return false, tmdb.ErrKeyEmpty
+	}
+	var found bool
+	err := b.db.View(func(txn *badger.Txn) error {
+		_, err := txn.Get(key)
+		if err != nil && err != badger.ErrKeyNotFound {
+			return err
+		}
+		found = (err != badger.ErrKeyNotFound)
+		return nil
+	})
+	return found, err
+}
+
+func (b *BadgerDB) Set(key, value []byte) error {
+	if len(key) == 0 {
+		return tmdb.ErrKeyEmpty
+	}
+	if value == nil {
+		return tmdb.ErrValueNil
+	}
+	return b.db.Update(func(txn *badger.Txn) error {
+		return txn.Set(key, value)
+	})
+}
+
+func withSync(db *badger.DB, err error) error {
+	if err != nil {
+		return err
+	}
+	return db.Sync()
+}
+
+func (b *BadgerDB) SetSync(key, value []byte) error {
+	return withSync(b.db, b.Set(key, value))
+}
+
+func (b *BadgerDB) Delete(key []byte) error {
+	if len(key) == 0 {
+		return tmdb.ErrKeyEmpty
+	}
+	return b.db.Update(func(txn *badger.Txn) error {
+		return txn.Delete(key)
+	})
+}
+
+func (b *BadgerDB) DeleteSync(key []byte) error {
+	return withSync(b.db, b.Delete(key))
+}
+
+func (b *BadgerDB) Close() error {
+	return b.db.Close()
+}
+
+func (b *BadgerDB) Print() error {
+	return nil
+}
+
+func (b *BadgerDB) iteratorOpts(start, end []byte, opts badger.IteratorOptions) (*badgerDBIterator, error) {
+	if (start != nil && len(start) == 0) || (end != nil && len(end) == 0) {
+		return nil, tmdb.ErrKeyEmpty
+	}
+	txn := b.db.NewTransaction(false)
+	iter := txn.NewIterator(opts)
+	iter.Rewind()
+	iter.Seek(start)
+	if opts.Reverse && iter.Valid() && bytes.Equal(iter.Item().Key(), start) {
+		// If we're going in reverse, our starting point was "end",
+		// which is exclusive.
+		iter.Next()
+	}
+	return &badgerDBIterator{
+		reverse: opts.Reverse,
+		start:   start,
+		end:     end,
+
+		txn:  txn,
+		iter: iter,
+	}, nil
+}
+
+func (b *BadgerDB) Iterator(start, end []byte) (tmdb.Iterator, error) {
+	opts := badger.DefaultIteratorOptions
+	return b.iteratorOpts(start, end, opts)
+}
+
+func (b *BadgerDB) ReverseIterator(start, end []byte) (tmdb.Iterator, error) {
+	opts := badger.DefaultIteratorOptions
+	opts.Reverse = true
+	return b.iteratorOpts(end, start, opts)
+}
+
+func (b *BadgerDB) Stats() map[string]string {
+	return nil
+}
+
+func (b *BadgerDB) NewBatch() tmdb.Batch {
+	wb := &badgerDBBatch{
+		db:         b.db,
+		wb:         b.db.NewWriteBatch(),
+		firstFlush: make(chan struct{}, 1),
+	}
+	wb.firstFlush <- struct{}{}
+	return wb
+}
+
+var _ tmdb.Batch = (*badgerDBBatch)(nil)
+
+type badgerDBBatch struct {
+	db *badger.DB
+	wb *badger.WriteBatch
+
+	// Calling db.Flush twice panics, so we must keep track of whether we've
+	// flushed already on our own. If Write can receive from the firstFlush
+	// channel, then it's the first and only Flush call we should do.
+	//
+	// Upstream bug report:
+	// https://github.com/dgraph-io/badger/issues/1394
+	firstFlush chan struct{}
```

Comment:
I would still look into removing this. But that can be done in a followup PR I think.

Reply:
Can look into it later tonight

Reply:
I think it would be a good idea to tackle this while at it. But IMO, we can also track this in a separate issue. For upstreaming this work to tendermint, this would have to be revisited as well imo.

Reply:
Will do this right away in a follow up PR.